libcasr/src/js.rs

```diff
@@ -0,0 +1,390 @@
+
+        /// Parse substring of `entry` related to location in some file
+        ///
+        /// # Arguments
+        ///
+        /// * `entry` - initial string for stacktrace entry
+        ///
+        /// * `stentry` - `StacktraceEntry` to save parsed location into
+        ///
+        /// * `loc` - substring to extract location information from
+        fn parse_location(entry: &str, stentry: &mut StacktraceEntry, loc: &str) -> Result<()> {
+            // filename[:line[:column]]
+            if loc.is_empty() {
+                return Err(Error::Casr(format!(
+                    "Couldn't parse location. Entry: {entry}"
+                )));
+            }
+
+            let mut debug: Vec<String> = loc.split(':').map(|s| s.to_string()).collect();
+            if debug.len() > 3 {
+                // Filename contains ':' so all the elements except
+                // the last 2 belong to filename
+                debug = loc.rsplitn(3, ':').map(|s| s.to_string()).collect();
+                debug.reverse();
+            }
+            if debug[0].contains("://") {
+                debug[0] = debug[0].rsplit("://").next().unwrap().to_string();
+            }
+            if debug[0].is_empty() || !debug[0].starts_with('/') {
+                // Filter empty filenames as related entries cannot provide enough info
+                // (e.g. "at func (:3:10)" cannot be attached to any source file)
+                // Filter non-absolute paths as they are related to internal JS modules
+                stentry.debug.file = "native".to_string();
+            } else {
+                stentry.debug.file = debug[0].to_string();
+            }
+            if debug.len() == 1 {
+                // Location contains filename only
+                return Ok(());
+            }
+
+            stentry.debug.line = if let Ok(line) = debug[1].parse::<u64>() {
+                line
+            } else {
+                return Err(Error::Casr(format!(
+                    "Couldn't parse line number {}. Entry: {entry}",
+                    debug[1]
+                )));
+            };
+            if debug.len() == 3 {
+                stentry.debug.column = if let Ok(column) = debug[2].parse::<u64>() {
+                    column
+                } else {
+                    return Err(Error::Casr(format!(
+                        "Couldn't parse column number {}. Entry: {entry}",
+                        debug[2]
+                    )));
+                }
+            }
+            Ok(())
+        }
+
+        /// Parse substring of `entry` related to location in some file
+        ///
+        /// # Arguments
+        ///
+        /// * `entry` - initial string for stacktrace entry
+        ///
+        /// * `stentry` - `StacktraceEntry` to save parsed location into
+        fn parse_eval(entry: &str, stentry: &mut StacktraceEntry) -> Result<()> {
```

Comment:
Why we need output mutable parameter? Maybe just return as `Result<StacktraceEntry>`?

Reply:
Done